src/aws/mod.rs

```diff
@@ -144,3 +151,159 @@ impl AsyncSleep for TokioSleep {
+            // `BytesSent` event to ensure that we capture the data flowing out as API calls.
+            let result = fut.await;
+            if let Ok(response) = &result {
+                let byte_size = maybe_bytes_sent
+                    .map(|s| s.load(Ordering::Acquire))
+                    .unwrap_or(0);
+
+                // TODO: Should we actually emit for any other range of status codes? Right now, `is_success` is true
+                // for `200 <= status < 300`, which feels comprehensive... but are there other valid statuses?
+                if response.http().status().is_success() && byte_size != 0 {
+                    emit!(AwsBytesSent {
+                        byte_size,
+                        region: Some(region),
+                        endpoint: endpoint.as_str(),
+                    });
+                }
+            }
+
+            result
+        })
+    }
+}
+
+struct BodyCaptureCallback {
+    bytes_sent: usize,
+    shared_bytes_sent: Arc<AtomicUsize>,
+}
+
+impl BodyCaptureCallback {
+    fn new() -> (Self, Arc<AtomicUsize>) {
+        let shared_bytes_sent = Arc::new(AtomicUsize::new(0));
+
+        (
+            Self {
+                bytes_sent: 0,
+                shared_bytes_sent: Arc::clone(&shared_bytes_sent),
+            },
+            shared_bytes_sent,
+        )
+    }
+}
+
+impl BodyCallback for BodyCaptureCallback {
+    fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
+        // This gets called every time a chunk is read from the request body, which includes both static chunks and
+        // streaming bodies. Just add the chunk's length to our running tally.
+        self.bytes_sent += bytes.len();
+        Ok(())
+    }
+
+    fn trailers(&self) -> Result<Option<headers::HeaderMap<headers::HeaderValue>>, BoxError> {
+        Ok(None)
+    }
+
+    fn make_new(&self) -> Box<dyn BodyCallback> {
+        // We technically don't use retries within the AWS side of the API clients, but we have to satisfy this trait
+        // method, because `aws_smithy_http` uses the retry layer from `tower`, which clones the request regardless
+        // before it even executes the first attempt... so there's no reason not to make it technically correct.
+        Box::new(Self {
+            bytes_sent: 0,
+            shared_bytes_sent: Arc::clone(&self.shared_bytes_sent),
+        })
+    }
+}
+
+impl Drop for BodyCaptureCallback {
+    fn drop(&mut self) {
+        // This is where we actually emit. We specifically emit here, and not in `trailers`, because despite the
+        // documentation that `trailers` is called after all chunks of the body are successfully read, `hyper` won't
+        // continue polling a body if it knows it's gotten all the available bytes i.e. it doesn't necessarily drive it
+        // until `poll_data` returns `None`. This means the only consistent place to know that the body is "done" is
+        // when it's dropped.
```

Comment:
Yep. Chatted with one of the engineers working on the AWS Rust SDK who said they'd be reworking this area of the code in a way that would allow the request body to be wrapped with a custom type on our end, so as to avoid needing the `BodyCallback` stuff at all, given that it can't function correctly as implemented, given what `hyper` does.